lezioni/2-Programmare in Java/challenge/Picred/File.java

```diff
@@ -0,0 +1,52 @@
+import java.io.BufferedReader;
+import java.io.BufferedWriter;
+import java.lang.StringBuilder;
+import java.io.FileWriter;
+import java.io.FileReader;
+import java.io.IOException;
+import java.lang.String;
+
+public class File{
+    private String file = "";
+
+
+    public File(String path){
+        this.file = path;
+    }
+
+
+    public String readFirstLine()  {
+        try (BufferedReader reader = new BufferedReader(new FileReader(file))) {
+            String line = reader.readLine();
+            return line;
+        } catch (IOException e) {
+            e.printStackTrace();
+            return "ERROR!";
+        }
+    }
+
+
+    public String readAllLines()  {
+        try (BufferedReader reader = new BufferedReader(new FileReader(file))) {
+            StringBuilder builder = new StringBuilder();
+            String line = "";
+            while((line = reader.readLine()) != null){
+                builder.append(line);
+            }
+            return builder.toString();
+        } catch (IOException e) {
+            e.printStackTrace();
+            return "File not found!";
```

Comment:
Come faccio a distinguere se non ho trovato il file o sto leggendo un file che ha una sola riga: "File not found"?
Lo stesso discorso in `readFirstLine`.

Reply:
Mh, forse non ho capito bene
1. In entrambe le funzioni, si arriva nel catch solo se il file non esiste, giusto?

2. Mentre, dimmi se ho capito bene, il secondo problema è: se chiamo `readAllLines()` e mi restituisce una sola riga, come faccio a sapere che quel file è effettivamente formato da una sola riga?

Reply:
Il problema è che la funzione restituisce la stringa "File not found!":

- Se si entra nel catch perché il file non esiste (o avviene qualche IOException)
- Se, leggendo il file correttamente, questo contiene solo la riga "File not found!"

Il client non ha modo per differenziare in quale dei due casi si trova

Lo stesso discorso vale per la funzione `readFirstLine`:

- Se si entra nel catch perché il file non esiste (o avviene qualche IOException)
- Se, leggendo il file correttamente, la prima riga è "ERROR!"

Ovviamente io, come utente, ho modo di vedere l'output prodotto da `e.printStackTrace();`, ma il programma dovrebbe poter fare la distinzione in maniera interna. Le due opzioni principali sarebbero

- restituire una stringa vuota
- restituire null nel caso si sia verificato un errore
- lasciare al client la gestione dell'eccezione

Trattandosi di un wrapper, probabilmente la prima potrebbe essere la più comoda, ma sono tutte e tre valide.